broadinstitute/single_cell_portal_core
# Enable cell filtering by all facets, not just top 5 (SCP-5361)

This lets users filter plotted cells by _all_ eligible categories, not just the top 5.  It vastly broadens search capability.

https://github.com/broadinstitute/single_cell_portal_core/assets/1334561/b641c7ba-ea08-488d-b7ad-ff800908311b

## Overview
Previously (#1897), only up to 5 facets (i.e. categories, annotations) were shown for cell filtering.  These initial facets were sorted by relevance, which helps surface interesting ways to search.  The initial set was limited to 5 facets to enable fast time-to-interactive (TTI), and also to not overwhelm the server.  However, only showing 5 out of often 25+ eligible facets severely narrowed users' search capabilities.  User frustration around that was foreseeable.

Now, all eligible facets are shown, and remaining facets after the initial load become interactive as data is retrieved.  To keep UX TTI fast and server saturation lower, data is fetched in batches / chunks of <= 5 facets at a time, rather than all remaining facets at once.  Even after 25+ facets are fully loaded, cell filtering is still instant in a real, representative dataset of 48K cells.

## Known issues
To get user feedback sooner, this enhancement achieves the main user-facing goal of quickly loading all facets, but ships with some rough edges.  Examples aimed for refinement in a subsequent PR for this ticket:

* If you click "Cell filtering" _while data is loading_, then unloaded facets won't become interactive while the panel is in view.  You need to exit the panel then re-enter for newly-loaded facets to become interactive.

* Too many requests are sent.  This is a pre-existing issue, but it's exacerbated in this enhancement.  Duplicative requests should be eliminated.

* Requests should be fully serialized.  The `/facets` requests sometimes begin in parallel, and sometimes _begin_ serially. 
 Each batched request _should not start_ until the previous one in the stream has completed.

These issues seem fine for an allow-listed beta, but warrant refinement before general availability.

## Test
To manually test:
1. Go to "Human milk - differential expression" with cell filtering enabled
2. Soon after the page loads, click "Cell filtering"
3. Confirm the first 5 facets appear and are interactive
4. Scroll fully down in panel, check if any facets are loaded
5. Confirm any loading facets appear grey, with loading icon, and filters disabled
6. If any facets are loading, click the "<-" to go back to "Options" panel
7. Wait a bit, click "Cell filtering" again
8. Confirm facets have finished loading (repeat 6-7 if so)
9. Scroll down to bottom of panel
10. Click any filter in "Antibiotics during delivery"
11. Confirm plot updates

This relates to SCP-5361.
